// @cloudflare/codemode/dist/ai.js
import { n as sanitizeToolName, t as generateTypes } from "./types-CpkEgXwN.js";
import { asSchema, tool } from "ai";
import { z } from "zod";
import * as acorn from "acorn";

//#region src/tool.ts
const DEFAULT_DESCRIPTION = `Execute code to achieve a goal.

Available:
{{types}}

Write an async arrow function in JavaScript that returns the result.
Do NOT use TypeScript syntax — no type annotations, interfaces, or generics.
Do NOT define named functions then call them — just write the arrow function body directly.

Example: async () => { const r = await codemode.searchWeb({ query: "test" }); return r; }`;
const codeSchema = z.object({ code: z.string().describe("JavaScript async arrow function to execute") });
function normalizeCode(code) {
	const trimmed = code.trim();
	if (!trimmed) return "async () => {}";
	try {
		const ast = acorn.parse(trimmed, {
			ecmaVersion: "latest",
			sourceType: "module"
		});
		if (ast.body.length === 1 && ast.body[0].type === "ExpressionStatement") {
			if (ast.body[0].expression.type === "ArrowFunctionExpression") return trimmed;
		}
		const last = ast.body[ast.body.length - 1];
		if (last?.type === "ExpressionStatement") {
			const exprStmt = last;
			return `async () => {\n${trimmed.slice(0, last.start)}return (${trimmed.slice(exprStmt.expression.start, exprStmt.expression.end)})\n}`;
		}
		return `async () => {\n${trimmed}\n}`;
	} catch {
		return `async () => {\n${trimmed}\n}`;
	}
}
/**
* Create a codemode tool that allows LLMs to write and execute code
* with access to your tools in a sandboxed environment.
*
* Returns an AI SDK compatible tool.
*/
function hasNeedsApproval(t) {
	return "needsApproval" in t && t.needsApproval != null;
}
function createCodeTool(options) {
	const tools = {};
	for (const [name, t] of Object.entries(options.tools)) if (!hasNeedsApproval(t)) tools[name] = t;
	const types = generateTypes(tools);
	const executor = options.executor;
	return tool({
		description: (options.description ?? DEFAULT_DESCRIPTION).replace("{{types}}", types),
		inputSchema: codeSchema,
		execute: async ({ code }) => {
			const fns = {};
			for (const [name, t] of Object.entries(tools)) {
				const execute = "execute" in t ? t.execute : void 0;
				if (execute) {
					const rawSchema = "inputSchema" in t ? t.inputSchema : "parameters" in t ? t.parameters : void 0;
					const schema = rawSchema != null ? asSchema(rawSchema) : void 0;
					fns[sanitizeToolName(name)] = schema?.validate ? async (args) => {
						const result = await schema.validate(args);
						if (!result.success) throw result.error;
						return execute(result.value);
					} : execute;
				}
			}
			const normalizedCode = normalizeCode(code);
			const executeResult = await executor.execute(normalizedCode, fns);
			if (executeResult.error) {
				const logCtx = executeResult.logs?.length ? `\n\nConsole output:\n${executeResult.logs.join("\n")}` : "";
				throw new Error(`Code execution failed: ${executeResult.error}${logCtx}`);
			}
			const output = {
				code,
				result: executeResult.result
			};
			if (executeResult.logs) output.logs = executeResult.logs;
			return output;
		}
	});
}

//#endregion
export { createCodeTool };
