// syntaqlite/dist/dialect.js
// Copyright 2025 The syntaqlite Authors. All rights reserved.
// Licensed under the Apache License, Version 2.0.
export const BUILTIN_PRESETS = [
    // SQLite is a side module; the runtime loads it via loadDynamicLibrary.
    { id: "sqlite", label: "SQLite", wasmUrl: "syntaqlite-sqlite.wasm", symbol: "syntaqlite_sqlite_dialect" },
];
export class DialectManager {
    constructor(config = {}) {
        this.active = undefined;
        this.customLabel = undefined;
        this.cache = new Map();
        this.presets = config.presets ?? BUILTIN_PRESETS;
        this.onDialectChanged = config.onDialectChanged;
        this.activePresetId = this.presets[0]?.id ?? "";
    }
    getPresets() {
        return this.presets;
    }
    async loadDefault(engine) {
        await this.selectPreset(engine, this.presets[0]);
    }
    async selectPreset(engine, preset) {
        if (!engine.ready) {
            engine.updateStatus("Runtime is not initialized yet.", true);
            this.onDialectChanged?.();
            return;
        }
        try {
            const binding = await this.loadCached(engine, preset.wasmUrl, preset.symbol, preset.label);
            engine.setDialectPointer(binding.ptr);
            this.active = binding;
            this.activePresetId = preset.id;
            this.customLabel = undefined;
            engine.updateStatus(`Dialect: ${preset.label}`);
        }
        catch (err) {
            engine.updateStatus(`Failed to load ${preset.label}: ${err.message}`, true);
        }
        this.onDialectChanged?.();
    }
    async loadFromFile(engine, file, symbol) {
        if (!engine.ready) {
            return "Runtime is not initialized yet.";
        }
        const url = URL.createObjectURL(file);
        try {
            const binding = await engine.loadDialectFromUrl(url, symbol);
            binding.label = file.name;
            engine.setDialectPointer(binding.ptr);
            this.active = binding;
            this.activePresetId = "custom";
            this.customLabel = file.name;
            engine.updateStatus(`Dialect: ${file.name}`);
            return undefined;
        }
        catch (err) {
            return err.message;
        }
        finally {
            URL.revokeObjectURL(url);
        }
    }
    async loadCached(engine, wasm, symbol, label) {
        const cached = this.cache.get(symbol);
        if (cached)
            return cached;
        const binding = await engine.loadDialectFromUrl(wasm, symbol);
        binding.label = label;
        this.cache.set(symbol, binding);
        return binding;
    }
}
