// syntaqlite/dist/engine.js
// Copyright 2025 The syntaqlite Authors. All rights reserved.
// Licensed under the Apache License, Version 2.0.
const DEFAULT_RUNTIME_JS = new URL("../wasm/syntaqlite-runtime.js", import.meta.url).href;
const DEFAULT_RUNTIME_WASM = new URL("../wasm/syntaqlite-runtime.wasm", import.meta.url).href;
export class Engine {
    constructor(config = {}) {
        this.status = "Loading...";
        this.statusError = false;
        this.module = undefined;
        this.encoder = new TextEncoder();
        this.decoder = new TextDecoder();
        this.setDialectRaw = undefined;
        this.clearDialectRaw = undefined;
        this.allocRaw = undefined;
        this.freeRaw = undefined;
        this.astRaw = undefined;
        this.astJsonRaw = undefined;
        this.fmtRaw = undefined;
        this.diagnosticsRaw = undefined;
        this.semanticTokensRaw = undefined;
        this.completionsRaw = undefined;
        this.resultPtrRaw = undefined;
        this.resultLenRaw = undefined;
        this.resultFreeRaw = undefined;
        this.setSqliteVersionRaw = undefined;
        this.setCflagRaw = undefined;
        this.clearCflagRaw = undefined;
        this.clearAllCflagsRaw = undefined;
        this.getCflagListRaw = undefined;
        this.setSessionContextRaw = undefined;
        this.clearSessionContextRaw = undefined;
        this.setSessionContextDdlRaw = undefined;
        this.setLanguageModeRaw = undefined;
        this.extractRaw = undefined;
        this.currentLangMode = "sql";
        /** Last session context applied, so it can be re-applied after dialect switches. */
        this.sessionContext = null;
        this.config = config;
    }
    get ready() {
        return this.module !== undefined;
    }
    updateStatus(text, isError = false) {
        this.status = text;
        this.statusError = isError;
    }
    async load() {
        const module = await loadRuntimeModule(this.config);
        this.module = module;
        this.setDialectRaw = this.tryResolveRuntimeFn("wasm_set_dialect");
        this.clearDialectRaw = this.tryResolveRuntimeFn("wasm_clear_dialect");
        this.allocRaw = this.resolveRuntimeFn("wasm_alloc");
        this.freeRaw = this.resolveRuntimeFn("wasm_free");
        this.astRaw = this.tryResolveRuntimeFn("wasm_ast");
        this.astJsonRaw = this.tryResolveRuntimeFn("wasm_ast_json");
        this.fmtRaw = this.resolveRuntimeFn("wasm_fmt");
        this.diagnosticsRaw = this.tryResolveRuntimeFn("wasm_diagnostics");
        this.semanticTokensRaw = this.tryResolveRuntimeFn("wasm_semantic_tokens");
        this.completionsRaw = this.tryResolveRuntimeFn("wasm_completions");
        this.resultPtrRaw = this.resolveRuntimeFn("wasm_result_ptr");
        this.resultLenRaw = this.resolveRuntimeFn("wasm_result_len");
        this.resultFreeRaw = this.resolveRuntimeFn("wasm_result_free");
        this.setSqliteVersionRaw = this.tryResolveRuntimeFn("wasm_set_sqlite_version");
        this.setCflagRaw = this.tryResolveRuntimeFn("wasm_set_cflag");
        this.clearCflagRaw = this.tryResolveRuntimeFn("wasm_clear_cflag");
        this.clearAllCflagsRaw = this.tryResolveRuntimeFn("wasm_clear_all_cflags");
        this.getCflagListRaw = this.tryResolveRuntimeFn("wasm_get_cflag_list");
        this.setSessionContextRaw = this.tryResolveRuntimeFn("wasm_set_session_context");
        this.clearSessionContextRaw = this.tryResolveRuntimeFn("wasm_clear_session_context");
        this.setSessionContextDdlRaw = this.tryResolveRuntimeFn("wasm_set_session_context_ddl");
        this.setLanguageModeRaw = this.tryResolveRuntimeFn("wasm_set_language_mode");
        this.extractRaw = this.tryResolveRuntimeFn("wasm_extract");
    }
    resolveRuntimeFn(symbol) {
        const fn = this.module[`_${symbol}`];
        if (typeof fn !== "function") {
            throw new Error(`missing runtime function: _${symbol}`);
        }
        return fn;
    }
    /** Like resolveRuntimeFn but returns undefined if not found. */
    tryResolveRuntimeFn(symbol) {
        const fn = this.module[`_${symbol}`];
        return typeof fn === "function" ? fn : undefined;
    }
    resolveDialectFn(symbol, localScope = undefined) {
        if (localScope && typeof localScope[symbol] === "function") {
            return localScope[symbol];
        }
        if (localScope && typeof localScope[`_${symbol}`] === "function") {
            return localScope[`_${symbol}`];
        }
        const direct = this.module[`_${symbol}`];
        if (typeof direct === "function") {
            return direct;
        }
        if (typeof this.module.cwrap === "function") {
            try {
                return this.module.cwrap(symbol, "number", []);
            }
            catch {
                // Fall through to explicit error below.
            }
        }
        throw new Error(`missing dialect symbol: ${symbol}`);
    }
    heapU8() {
        const heap = this.module.HEAPU8 || window.HEAPU8;
        if (!heap)
            throw new Error("runtime HEAPU8 is not available");
        return heap;
    }
    async loadDialectFromUrl(url, symbol) {
        const localScope = {};
        if (url) {
            if (typeof this.module.loadDynamicLibrary !== "function") {
                throw new Error("runtime module does not expose loadDynamicLibrary");
            }
            const maybePromise = this.module.loadDynamicLibrary(url, { loadAsync: true, global: false, nodelete: true }, localScope);
            if (maybePromise && typeof maybePromise.then === "function") {
                await maybePromise;
            }
        }
        let ptr;
        try {
            const fn = this.resolveDialectFn(symbol, localScope);
            ptr = fn() >>> 0;
        }
        catch {
            throw new Error(`Symbol "${symbol}" not found in the WASM module.`);
        }
        if (ptr === 0)
            throw new Error(`Symbol "${symbol}" returned undefined.`);
        this.setDialectPointer(ptr);
        return { symbol, ptr, label: symbol };
    }
    withInput(sql, fn) {
        const input = this.encoder.encode(sql);
        const ptr = this.allocRaw(input.length);
        if (input.length > 0 && ptr === 0)
            throw new Error("allocation failed");
        if (input.length > 0)
            this.heapU8().set(input, ptr);
        try {
            return fn(ptr, input.length);
        }
        finally {
            this.freeRaw(ptr, input.length);
        }
    }
    readAndClearResult() {
        const ptr = this.resultPtrRaw();
        const len = this.resultLenRaw();
        const text = len === 0 ? "" : this.decoder.decode(this.heapU8().subarray(ptr, ptr + len));
        this.resultFreeRaw();
        return text;
    }
    setDialectPointer(ptr) {
        if (!this.setDialectRaw)
            throw new Error("dialect switching not supported by this runtime");
        const status = this.setDialectRaw(ptr >>> 0);
        const detail = this.readAndClearResult();
        if (status !== 0) {
            throw new Error(detail || `wasm_set_dialect failed with status ${status}`);
        }
        // The WASM invalidates the LSP host on dialect switch, discarding any
        // session context. Re-apply it so callers don't have to track this.
        this.reapplySessionContext();
    }
    reapplySessionContext() {
        if (!this.sessionContext)
            return;
        if (this.sessionContext.kind === "json") {
            this.applySessionContextJson(this.sessionContext.json);
        }
        else {
            this.applySessionContextDdl(this.sessionContext.sql);
        }
    }
    clearDialectPointer() {
        if (!this.clearDialectRaw)
            return;
        this.clearDialectRaw();
        this.readAndClearResult();
    }
    runAst(sql) {
        if (!this.astRaw)
            return { ok: false, text: "AST dump not supported by this runtime" };
        const status = this.withInput(sql, (ptr, len) => this.astRaw(ptr, len));
        const text = this.readAndClearResult();
        return { ok: status === 0, text };
    }
    runAstJson(sql) {
        if (!this.astJsonRaw)
            return { ok: false, error: "AST JSON not supported by this runtime" };
        const count = this.withInput(sql, (ptr, len) => this.astJsonRaw(ptr, len));
        const text = this.readAndClearResult();
        if (count < 0)
            return { ok: false, error: text };
        if (count === 0)
            return { ok: true, statements: [] };
        try {
            return { ok: true, statements: JSON.parse(text, (_, v) => (v === null ? undefined : v)) };
        }
        catch (e) {
            return { ok: false, error: `JSON parse error: ${e.message}` };
        }
    }
    runFmt(sql, opts) {
        const status = this.withInput(sql, (ptr, len) => this.fmtRaw(ptr, len, opts.lineWidth, opts.indentWidth, opts.keywordCase, opts.semicolons ? 1 : 0));
        const text = this.readAndClearResult();
        return { ok: status === 0, text };
    }
    /** Run semantic token analysis over a byte range. Returns a pre-encoded
     *  Uint32Array (5 u32s per token: deltaLine, deltaStartChar, length,
     *  legendIndex, 0) ready for Monaco, or undefined on failure.
     *  Pass rangeStart=0 and rangeEnd=0xFFFFFFFF for the full document. */
    runSemanticTokens(sql, rangeStart = 0, rangeEnd = 0xffffffff, version = 1) {
        if (!this.semanticTokensRaw)
            return undefined;
        try {
            const count = this.withInput(sql, (ptr, len) => this.semanticTokensRaw(ptr, len, rangeStart, rangeEnd, version));
            if (count <= 0) {
                this.resultFreeRaw();
                return count === 0 ? new Uint32Array(0) : undefined;
            }
            // Read raw bytes from RESULT_BUF as a Uint32Array (5 u32s per token).
            const rptr = this.resultPtrRaw();
            const rlen = this.resultLenRaw();
            const bytes = this.heapU8().slice(rptr, rptr + rlen);
            this.resultFreeRaw();
            return new Uint32Array(bytes.buffer, bytes.byteOffset, bytes.byteLength / 4);
        }
        catch (e) {
            console.warn("wasm_semantic_tokens failed:", e);
            return undefined;
        }
    }
    runDiagnostics(sql, version = 1) {
        if (!this.diagnosticsRaw)
            return { ok: false, diagnostics: [] };
        try {
            const count = this.withInput(sql, (ptr, len) => this.diagnosticsRaw(ptr, len, version));
            const text = this.readAndClearResult();
            if (count < 0)
                return { ok: false, diagnostics: [] };
            if (count === 0)
                return { ok: true, diagnostics: [] };
            const diagnostics = JSON.parse(text);
            return { ok: true, diagnostics };
        }
        catch (e) {
            console.warn("wasm_diagnostics failed:", e);
            return { ok: false, diagnostics: [] };
        }
    }
    runCompletions(sql, offset, version = 1) {
        if (!this.completionsRaw)
            return { ok: false, items: [] };
        try {
            const count = this.withInput(sql, (ptr, len) => this.completionsRaw(ptr, len, offset >>> 0, version));
            const text = this.readAndClearResult();
            if (count < 0)
                return { ok: false, items: [] };
            if (count === 0)
                return { ok: true, items: [] };
            const items = JSON.parse(text);
            return { ok: true, items };
        }
        catch (e) {
            console.warn("wasm_completions failed:", e);
            return { ok: false, items: [] };
        }
    }
    /** Set the active language mode. Must be called before running diagnostics or semantic
     *  tokens so the WASM can dispatch to the correct implementation automatically.
     *  @experimental Embedded language support is experimental and may change. */
    setLanguageMode(lang) {
        this.currentLangMode = lang;
        if (!this.setLanguageModeRaw)
            return;
        const code = lang === "sql" ? 0xFFFFFFFF : (lang === "python" ? 0 : 1);
        this.setLanguageModeRaw(code);
    }
    /** Extract SQL fragments from `source`. Returns empty in SQL mode (O(1) fast path).
     *  In embedded mode the WASM extractor runs based on the language set by setLanguageMode.
     *  @experimental Embedded language support is experimental and may change. */
    runExtract(source) {
        if (this.currentLangMode === "sql")
            return { ok: true, fragments: [] };
        if (!this.extractRaw)
            return { ok: true, fragments: [] };
        try {
            const count = this.withInput(source, (ptr, len) => this.extractRaw(ptr, len));
            const text = this.readAndClearResult();
            if (count < 0)
                return { ok: false, fragments: [] };
            if (count === 0)
                return { ok: true, fragments: [] };
            const fragments = JSON.parse(text);
            return { ok: true, fragments };
        }
        catch (e) {
            console.warn("wasm_extract failed:", e);
            return { ok: false, fragments: [] };
        }
    }
    setSqliteVersion(version) {
        if (!this.setSqliteVersionRaw)
            return;
        const status = this.withInput(version, (ptr, len) => this.setSqliteVersionRaw(ptr, len));
        const detail = this.readAndClearResult();
        if (status !== 0) {
            throw new Error(detail || `wasm_set_sqlite_version failed with status ${status}`);
        }
    }
    setCflag(name) {
        if (!this.setCflagRaw)
            return;
        const status = this.withInput(name, (ptr, len) => this.setCflagRaw(ptr, len));
        const detail = this.readAndClearResult();
        if (status !== 0) {
            throw new Error(detail || `wasm_set_cflag failed with status ${status}`);
        }
    }
    clearCflag(name) {
        if (!this.clearCflagRaw)
            return;
        const status = this.withInput(name, (ptr, len) => this.clearCflagRaw(ptr, len));
        const detail = this.readAndClearResult();
        if (status !== 0) {
            throw new Error(detail || `wasm_clear_cflag failed with status ${status}`);
        }
    }
    clearAllCflags() {
        if (!this.clearAllCflagsRaw)
            return;
        this.clearAllCflagsRaw();
    }
    getCflagList() {
        if (!this.getCflagListRaw)
            return [];
        this.getCflagListRaw();
        const text = this.readAndClearResult();
        if (!text)
            return [];
        try {
            return JSON.parse(text);
        }
        catch {
            return [];
        }
    }
    setSessionContext(json) {
        this.sessionContext = { kind: "json", json };
        this.applySessionContextJson(json);
    }
    clearSessionContext() {
        this.sessionContext = null;
        if (!this.clearSessionContextRaw)
            return;
        this.clearSessionContextRaw();
    }
    setSessionContextDdl(sql) {
        const result = this.applySessionContextDdl(sql);
        if (result.ok)
            this.sessionContext = { kind: "ddl", sql };
        return result;
    }
    applySessionContextJson(json) {
        if (!this.setSessionContextRaw)
            return;
        const status = this.withInput(json, (ptr, len) => this.setSessionContextRaw(ptr, len));
        const detail = this.readAndClearResult();
        if (status !== 0) {
            throw new Error(detail || `wasm_set_session_context failed with status ${status}`);
        }
    }
    applySessionContextDdl(sql) {
        if (!this.setSessionContextDdlRaw)
            return { ok: false, error: "DDL context not supported" };
        const status = this.withInput(sql, (ptr, len) => this.setSessionContextDdlRaw(ptr, len));
        const detail = this.readAndClearResult();
        if (status !== 0)
            return { ok: false, error: detail || "DDL parse failed" };
        return { ok: true };
    }
}
function loadRuntimeModule(config) {
    return new Promise((resolve, reject) => {
        const jsPath = config.runtimeJsPath ?? DEFAULT_RUNTIME_JS;
        const wasmPath = config.runtimeWasmPath ?? (config.runtimeJsPath ? config.runtimeJsPath.replace(/\.js$/, ".wasm") : DEFAULT_RUNTIME_WASM);
        const moduleConfig = {
            noInitialRun: true,
            locateFile(path) {
                if (path === "syntaqlite_wasm.wasm" || path === "syntaqlite-wasm.wasm") {
                    return wasmPath;
                }
                return path;
            },
            onRuntimeInitialized() {
                resolve(moduleConfig);
            },
            onAbort(reason) {
                reject(new Error(`runtime aborted: ${reason}`));
            },
        };
        window.Module = moduleConfig;
        const script = document.createElement("script");
        script.src = jsPath;
        script.async = true;
        script.onerror = () => reject(new Error(`failed to load ${jsPath}`));
        document.head.appendChild(script);
    });
}
